src/test/java/com/github/packageurl/PackageURLBuilderTest.java

```diff
@@ -21,76 +21,68 @@
-                .build();
-
-        assertEquals("pkg:generic/namespace/name@version?key_1.1-=value#subpath", purl.toString());
-
-        purl = PackageURLBuilder.aPackageURL()
-                .withType(PackageURL.StandardTypes.GENERIC)
-                .withNamespace("/////")
-                .withName("name")
-                .withVersion("version")
-                .withQualifier("key","value")
-                .withSubpath("/////")
-                .build();
-
-        assertEquals("pkg:generic/name@version?key=value", purl.toString());
-
-        purl = PackageURLBuilder.aPackageURL()
-                .withType(PackageURL.StandardTypes.GENERIC)
-                .withNamespace("")
-                .withName("name")
-                .withVersion("version")
-                .withQualifier("key","value")
-                .withQualifier("next","value")
-                .withSubpath("")
-                .build();
+    @ParameterizedTest
+    @JsonClasspathSource({"test-suite-data.json", "components-constructor-only.json"})
+    void packageURLBuilder(JsonTestData json) throws MalformedPackageURLException {
+        if (json.getType() == null || json.getName() == null) {
+            assertTrue(json.isInvalid(), "valid test case with type or name `null`");
+            return;
+        }
+        PackageURLBuilder builder = aPackageURL()
+                .withType(json.getType())
+                .withName(json.getName());
+        String namespace = json.getNamespace();
+        if (namespace != null) {
+            builder.withNamespace(namespace);
+        }
+        String version = json.getVersion();
+        if (version != null) {
+            builder.withVersion(version);
+        }
+        json.getQualifiers().forEach(builder::withQualifier);
+        String subpath = json.getSubpath();
+        if (subpath != null) {
+            builder.withSubpath(subpath);
+        }
+        if (json.isInvalid()) {
+            assertThrows(MalformedPackageURLException.class, builder::build);
+        } else {
+            assertBuilderEquals(json, builder);
+            PackageURL actual = builder.build();
+            assertEquals(json.getCanonicalPurl(), actual.canonicalize(), "canonical PURL");
+        }
+    }
 
-        assertEquals("pkg:generic/name@version?key=value&next=value", purl.toString());
+    private static void assertBuilderEquals(JsonTestData expected, PackageURLBuilder actual) {
+        assertEquals(expected.getType(), actual.getType(), "type");
+        assertEquals(expected.getNamespace(), actual.getNamespace(), "namespace");
+        assertEquals(expected.getName(), actual.getName(), "name");
+        assertEquals(expected.getVersion(), actual.getVersion(), "version");
+        assertEquals(expected.getSubpath(), actual.getSubpath(), "subpath");
+        assertNotNull(actual.getQualifiers(), "qualifiers");
+        assertThat(actual.getQualifiers()).containsExactlyInAnyOrderEntriesOf(expected.getQualifiers());
```

Comment:
I personally prefer AssertJ to JUnit, but I am not sure if we want the extra dependency, but if so, then there definitely shouldn't be a mix of both types of assertions, I think.

Anyway,

```java
    assertEquals(expected.getQualifiers(), actual.getQualifiers());
```

should work just fine. Why it just didn't use this initially, I don't know. Maybe it had to do with the handling of `null` inside the JSON data.